src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs

```diff
@@ -3,49 +3,217 @@
-            string key = reader.GetString()!;
-            Debug.Assert(metadata == MetadataPropertyName.Id || metadata == MetadataPropertyName.Ref);
+                string id = reader.GetString()!;
+                state.Current.MetadataId = id;
 
-            if (metadata == MetadataPropertyName.Id)
+                // Clear the MetadataPropertyName since we are done processing Id.
+                state.Current.MetadataPropertyName = MetadataPropertyName.NoMetadata;
+
+                if (converter.ClassType == ClassType.Enumerable && converter.CanHaveValuesMetadata)
+                {
+                    // Need to Read $values property name.
+                    state.Current.ObjectState = StackFrameObjectState.MetadataValuesPropertyName;
+                }
+                else
+                {
+                    // We are done reading metadata.
+                    state.Current.ObjectState = StackFrameObjectState.MetadataPropertyValue;
+                    value = default!;
+                    return true;
+                }
+            }
+
+            if (state.Current.ObjectState == StackFrameObjectState.MetadataRefPropertyEndObject)
             {
-                // Special case for dictionary properties since those do not push into the ReadStack.
-                // There is no need to check for enumerables since those will always be wrapped into JsonPreservableArrayReference<T> which turns enumerables into objects.
-                object value = state.Current.IsProcessingProperty(ClassType.Dictionary) ?
-                    state.Current.JsonPropertyInfo!.GetValueAsObject(state.Current.ReturnValue)! :
-                    state.Current.ReturnValue!;
+                if (!reader.Read())
+                {
+                    value = default!;
+                    return false;
+                }
 
-                state.ReferenceResolver.AddReferenceOnDeserialize(key, value);
+                if (reader.TokenType != JsonTokenType.EndObject)
+                {
+                    ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(converter.TypeToConvert);
+                }
+
+                // Clear the MetadataPropertyName since we are done processing Ref.
+                state.Current.MetadataPropertyName = MetadataPropertyName.NoMetadata;
+
+                value = (T)state.Current.ReturnValue!;
+                return true;
+            }
+
+            if (state.Current.ObjectState == StackFrameObjectState.MetadataValuesPropertyName)
+            {
+                if (!reader.Read())
+                {
+                    value = default!;
+                    return false;
+                }
+
+                if (reader.TokenType != JsonTokenType.PropertyName)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataPreservedArrayValuesNotFound(converter.TypeToConvert);
+                }
+
+                if (reader.GetString() != "$values")
```

Comment:
Thanks. Changing

Reply:
> Also, `GetString` returns the unescaped property name, we must compare against the raw bytes.

A test should have failed because of this. We should add one to make sure someone doesn't accidentally change it to `reader.GetString()` in the future.

Reply:
Yes @Jozkee is currently working on updating \ adding various tests and this should be one of them.